source/mqtt311_listener.c

```diff
@@ -0,0 +1,291 @@
+
+    listener->allocator = allocator;
+    listener->config = *config;
+
+    aws_mqtt_client_connection_acquire(config->connection);
+    aws_ref_count_init(&listener->ref_count, listener, s_aws_mqtt311_listener_on_zero_ref_count);
+
+    aws_task_init(
+        &listener->initialize_task, s_mqtt311_listener_initialize_task_fn, listener, "Mqtt311ListenerInitialize");
+    aws_task_init(
+        &listener->terminate_task, s_mqtt311_listener_terminate_task_fn, listener, "Mqtt311ListenerTerminate");
+
+    aws_mqtt311_listener_acquire(listener);
+    aws_event_loop_schedule_task_now(
+        s_mqtt_client_connection_get_event_loop(config->connection), &listener->initialize_task);
+
+    return listener;
+}
+
+struct aws_mqtt311_listener *aws_mqtt311_listener_acquire(struct aws_mqtt311_listener *listener) {
+    if (listener != NULL) {
+        aws_ref_count_acquire(&listener->ref_count);
+    }
+
+    return listener;
+}
+
+struct aws_mqtt311_listener *aws_mqtt311_listener_release(struct aws_mqtt311_listener *listener) {
+    if (listener != NULL) {
+        aws_ref_count_release(&listener->ref_count);
+    }
+
+    return NULL;
+}
+
+struct aws_mqtt311_callback_set_entry {
+    struct aws_allocator *allocator;
+
+    struct aws_linked_list_node node;
+
+    uint64_t id;
+
+    struct aws_mqtt311_callback_set callbacks;
+};
+
+void aws_mqtt311_callback_set_manager_init(
+    struct aws_mqtt311_callback_set_manager *manager,
+    struct aws_allocator *allocator,
+    struct aws_mqtt_client_connection *connection) {
+
+    manager->allocator = allocator;
+    manager->connection = connection; /* no need to ref count, this is assumed to be owned by the client connection */
+    manager->next_callback_set_entry_id = 1;
+
+    aws_linked_list_init(&manager->callback_set_entries);
+}
+
+void aws_mqtt311_callback_set_manager_clean_up(struct aws_mqtt311_callback_set_manager *manager) {
+    struct aws_linked_list_node *node = aws_linked_list_begin(&manager->callback_set_entries);
+    while (node != aws_linked_list_end(&manager->callback_set_entries)) {
+        struct aws_mqtt311_callback_set_entry *entry =
+            AWS_CONTAINER_OF(node, struct aws_mqtt311_callback_set_entry, node);
+        node = aws_linked_list_next(node);
+
+        aws_linked_list_remove(&entry->node);
+        aws_mem_release(entry->allocator, entry);
+    }
+}
+
+static struct aws_mqtt311_callback_set_entry *s_new_311_callback_set_entry(
+    struct aws_mqtt311_callback_set_manager *manager,
+    struct aws_mqtt311_callback_set *callback_set) {
+    struct aws_mqtt311_callback_set_entry *entry =
+        aws_mem_calloc(manager->allocator, 1, sizeof(struct aws_mqtt311_callback_set_entry));
+
+    entry->allocator = manager->allocator;
+    entry->id = manager->next_callback_set_entry_id++;
+    entry->callbacks = *callback_set;
+
+    AWS_LOGF_INFO(
+        AWS_LS_MQTT_GENERAL,
+        "id=%p: MQTT311 callback manager created new entry :%" PRIu64,
```

Comment:
Trivial: Can we add "id" to line up with the log below? "created new entry with id=%"